service/src/main/java/bio/terra/pipelines/common/utils/PipelineRunFilterSpecification.java

```diff
@@ -0,0 +1,112 @@
+package bio.terra.pipelines.common.utils;
+
+import bio.terra.pipelines.db.entities.Pipeline;
+import bio.terra.pipelines.db.entities.PipelineRun;
+import bio.terra.pipelines.service.exception.InvalidFilterException;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.Join;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.springframework.data.jpa.domain.Specification;
+
+/** Utility class for building dynamic JPA Specifications for filtering PipelineRun entities */
+public class PipelineRunFilterSpecification {
+
+  public static final String FILTER_STATUS = "status";
+  public static final String FILTER_JOB_ID = "jobId";
+  public static final String FILTER_PIPELINE_NAME = "pipelineName";
+  public static final String FILTER_DESCRIPTION = "description";
+
+  private PipelineRunFilterSpecification() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  /**
+   * Build a dynamic JPA Specification based on the provided map of user-specific filters
+   *
+   * @param filters Map of field names to filter values
+   * @param userId The user ID to filter by (always required)
+   * @return A Specification for filtering PipelineRun entities
+   */
+  public static Specification<PipelineRun> buildFilterSpecificationWithUserId(
+      Map<String, String> filters, String userId) {
+    return (root, query, criteriaBuilder) -> {
+      List<Predicate> predicates = new ArrayList<>();
+
+      // Always filter by userId
+      predicates.add(criteriaBuilder.equal(root.get("userId"), userId));
+
+      if (filters != null && !filters.isEmpty()) {
+        filters.forEach(
+            (key, value) -> {
+              if (value != null && !value.isEmpty()) {
+                switch (key) {
+                  case FILTER_STATUS:
```

Comment:
can we double check all of these columns are indexed in the db.  Description is prob the only one it would make sense to not index.

Reply:
userId and jobId are indexed, status is not.

would a status index be super useful here? I _think_ since we always filter on user_id, postgres is able to optimize the query by first leaning into the user_id index. so the unindexed status filtering would be on a fairly small amount of data. but if we expect users to have tons of jobs i could definitely buy into adding some type of index